### PR TITLE
Fix LaTeX error due to labels that become empty in simplify_labels

### LINKED ISSUE #8004: TickLabels.simplify_labels corrupts LaTeX label strings on Mollweide axes
This example script that makes a Mollweide projection plot with WCSAxes with matplotlib's `text.usetex` rcparam set to `True` crashes. By setting breakpoints, I found that it is the declination axis that is getting corrupted, and it is happening in `TickLabels.simplify_labels`.

## test.py

```python
#!/usr/bin/env python
import matplotlib
matplotlib.rcParams['text.usetex'] = True
from astropy import units as u
from astropy.visualization.wcsaxes.frame import EllipticalFrame
from astropy.wcs import WCS
from matplotlib import pyplot as plt
import numpy as np

header = {
    'NAXIS': 2,
    'NAXIS1': 360,
    'NAXIS2': 180,
    'CRPIX1': 180.5,
    'CRPIX2': 90.5,
    'CRVAL1': 180.0,
    'CRVAL2': 0.0,
    'CDELT1': -2 * np.sqrt(2) / np.pi,
    'CDELT2': 2 * np.sqrt(2) / np.pi,
    'CTYPE1': 'RA---MOL',
    'CTYPE2': 'DEC--MOL',
    'RADESYS': 'ICRS'}

wcs = WCS(header)
ax = plt.axes(frame_class=EllipticalFrame, projection=wcs)
ax.set_xlim(-0.5, header['NAXIS1'] - 0.5)
ax.set_ylim(-0.5, header['NAXIS2'] - 0.5)
ax.coords[0].set_ticks(exclude_overlapping=True, spacing=45 * u.deg)
ax.coords[1].set_ticks(exclude_overlapping=True, spacing=30 * u.deg)
ax.grid()
plt.savefig('test.png')
```

## Output

```
$ ./test.py 
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/texmanager.py", line 298, in _run_checked_subprocess
    stderr=subprocess.STDOUT)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 389, in check_output
    **kwargs).stdout
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 481, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['latex', '-interaction=nonstopmode', '--halt-on-error', '/Users/lpsinger/.matplotlib/tex.cache/3a1d4a91ea3bce281dc11a6a0700d7e5.tex']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./test.py", line 31, in <module>
    plt.savefig('test.png')
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/pyplot.py", line 688, in savefig
    res = fig.savefig(*args, **kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/figure.py", line 2097, in savefig
    self.canvas.print_figure(fname, **kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/backend_bases.py", line 2075, in print_figure
    **kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/backends/backend_agg.py", line 510, in print_png
    FigureCanvasAgg.draw(self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/backends/backend_agg.py", line 402, in draw
    self.figure.draw(self.renderer)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/artist.py", line 50, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/figure.py", line 1652, in draw
    renderer, self, artists, self.suppressComposite)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/image.py", line 138, in _draw_list_compositing_images
    a.draw(renderer)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/astropy/visualization/wcsaxes/core.py", line 392, in draw
    super().draw(renderer, inframe=inframe)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/artist.py", line 50, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/axes/_base.py", line 2604, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/image.py", line 138, in _draw_list_compositing_images
    a.draw(renderer)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/astropy/visualization/wcsaxes/core.py", line 48, in draw
    self.axes.draw_wcsaxes(renderer)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/astropy/visualization/wcsaxes/core.py", line 356, in draw_wcsaxes
    ticks_locs=ticks_locs[coord])
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/astropy/visualization/wcsaxes/coordinate_helpers.py", line 483, in _draw_ticks
    ticklabels_bbox=ticklabels_bbox)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/astropy/visualization/wcsaxes/ticklabels.py", line 157, in draw
    bb = super().get_window_extent(renderer)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/text.py", line 929, in get_window_extent
    bbox, info, descent = self._get_layout(self._renderer)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/text.py", line 313, in _get_layout
    ismath=ismath)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/backends/backend_agg.py", line 209, in get_text_width_height_descent
    s, fontsize, renderer=self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/texmanager.py", line 464, in get_text_width_height_descent
    dvifile = self.make_dvi(tex, fontsize)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/texmanager.py", line 328, in make_dvi
    texfile], tex)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/texmanager.py", line 307, in _run_checked_subprocess
    exc=exc.output.decode('utf-8')))
RuntimeError: latex was not able to process the following string:
b'$'

Here is the full report generated by latex:
This is pdfTeX, Version 3.14159265-2.6-1.40.19 (TeX Live 2018/MacPorts 2018.47642_7) (preloaded format=latex)
 restricted \write18 enabled.
entering extended mode
(/Users/lpsinger/.matplotlib/tex.cache/3a1d4a91ea3bce281dc11a6a0700d7e5.tex
LaTeX2e <2018-04-01> patch level 2
Babel <3.18> and hyphenation patterns for 46 language(s) loaded.
(/opt/local/share/texmf-texlive/tex/latex/base/article.cls
Document Class: article 2014/09/29 v1.4h Standard LaTeX document class
(/opt/local/share/texmf-texlive/tex/latex/base/size10.clo))
(/opt/local/share/texmf-texlive/tex/latex/type1cm/type1cm.sty)
(/opt/local/share/texmf-texlive/tex/latex/base/textcomp.sty
(/opt/local/share/texmf-texlive/tex/latex/base/ts1enc.def
(/opt/local/share/texmf-texlive/tex/latex/base/ts1enc.dfu)))
(/opt/local/share/texmf-texlive/tex/latex/base/inputenc.sty)
(/opt/local/share/texmf-texlive/tex/latex/geometry/geometry.sty
(/opt/local/share/texmf-texlive/tex/latex/graphics/keyval.sty)
(/opt/local/share/texmf-texlive/tex/generic/oberdiek/ifpdf.sty)
(/opt/local/share/texmf-texlive/tex/generic/oberdiek/ifvtex.sty)
(/opt/local/share/texmf-texlive/tex/generic/ifxetex/ifxetex.sty)

Package geometry Warning: Over-specification in `h'-direction.
    `width' (5058.9pt) is ignored.


Package geometry Warning: Over-specification in `v'-direction.
    `height' (5058.9pt) is ignored.

) (./3a1d4a91ea3bce281dc11a6a0700d7e5.aux)
(/opt/local/share/texmf-texlive/tex/latex/base/ts1cmr.fd)
*geometry* driver: auto-detecting
*geometry* detected driver: dvips
! Extra }, or forgotten $.
<recently read> }
                 
l.14 \fontsize{10.000000}{12.500000}{\sffamily $}
                                                 
No pages of output.
Transcript written on 3a1d4a91ea3bce281dc11a6a0700d7e5.log.
```

### PR DESCRIPTION
**Pull request by lpsinger**
Fixes #8004.

### CODE REVIEW

**Comment by pllim** [comment]
Thanks, @lpsinger ! Any chance we can add a test for this use case too?

**Comment by lpsinger** [comment]
I added a unit test, but it is failing with this error message:

```
RuntimeError: Failed to process string with tex because latex could not be found
```

Are there any CI jobs that run in an environment with LaTeX installed? Is there already a decorator or a pytest mark that I can use to skip this test if LaTeX is not installed?

**Comment by pllim** [comment]
Hmm. Not that I know of. You can try using `pytest.importorskip` feature to skip it.

**Comment by lpsinger** [comment]
I am using `checkdep_usetex` from Matplotlib now. See this example:

https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_usetex.py#L11-L12

**Comment by pllim** [comment]
Might need to skip it regardless if matplotlib version is too old. The job with Matplotlib 2.1.2 is failing. FYI.

**Comment by lpsinger** [comment]
Done.

**Comment by lpsinger** [comment]
Now it's failing with:

```
NameError: name 'TEX_MISSING' is not defined
```

I've no idea why, since it is not having a problem  with `MATPLOTLIB_LT_21`.

File: astropy/visualization/wcsaxes/tests/test_misc.py [hunk 1]
```diff
@@ -446,3 +449,36 @@ def test_time_wcs(time_spectral_wcs_2d):
     # with a time axis.
 
     plt.subplot(projection=time_spectral_wcs_2d)
+
+
+@pytest.mark.skipif('TEX_MISSING or MATPLOTLIB_LT_21')
```
Comment (pllim) [inline astropy/visualization/wcsaxes/tests/test_misc.py]: It is not clear to me why this failed either. What if we lose the quotes?
```suggestion
@pytest.mark.skipif(TEX_MISSING or MATPLOTLIB_LT_21)
```

**Comment by lpsinger** [comment]
I don't think that it's even respecting `@pytest.mark.skipif('MATPLOTLIB_LT_21')`. Something very strange and counterintuitive is going on. I use `pytest.mark.skipif` all the time, and this behavior does not make sense to me. Help!

**Comment by pllim** [comment]
I wonder if it is incompatible with `@ignore_matplotlibrc`. I will try to push a commit to your PR and see. I hope you don't mind.

File: astropy/visualization/wcsaxes/tests/test_misc.py [hunk 2]
```diff
@@ -446,3 +447,43 @@ def test_time_wcs(time_spectral_wcs_2d):
...
+        subplot_kw=dict(frame_class=EllipticalFrame, projection=wcs))
+    ax.set_xlim(-0.5, header['NAXIS1'] - 0.5)
+    ax.set_ylim(-0.5, header['NAXIS2'] - 0.5)
+    ax.coords[0].set_ticklabel(exclude_overlapping=True)
+    ax.coords[1].set_ticklabel(exclude_overlapping=True)
+    ax.coords[0].set_ticks(spacing=45 * u.deg)
+    ax.coords[1].set_ticks(spacing=30 * u.deg)
+    ax.grid()
+
+    if not matplotlib.checkdep_usetex(True):
```
Comment (pllim) [inline astropy/visualization/wcsaxes/tests/test_misc.py]: Why is this not at the top of the test function?

**Comment by pllim** [comment]
Re: pep8speaks -- My bad. Must have accidentally included it when I edited this on GitHub. I'll help you clean things up before merge, if you need the help. Sorry!

File: astropy/visualization/wcsaxes/tests/test_misc.py (hunk 2 above)
Comment (lpsinger) [reply astropy/visualization/wcsaxes/tests/test_misc.py]: Because I was flailing while trying to fix the skipif stuff.

**Comment by lpsinger** [comment]
> I will try to push a commit to your PR and see. I hope you don't mind.

Not at all. Please go ahead.

File: astropy/visualization/wcsaxes/tests/test_misc.py [hunk 3]
```diff
@@ -446,3 +448,40 @@ def test_time_wcs(time_spectral_wcs_2d):
     # with a time axis.
 
     plt.subplot(projection=time_spectral_wcs_2d)
+
+
+@ignore_matplotlibrc
+def test_simplify_labels_usetex(tmpdir):
+    """Regression test for https://github.com/astropy/astropy/issues/8004."""
+
+    if MATPLOTLIB_LT_22 or not matplotlib.checkdep_usetex(True):
```
Comment (pllim) [inline astropy/visualization/wcsaxes/tests/test_misc.py]: Turns out the failing version was 2.1.2, so technically it is `>=2.1`. That is why it wasn't working.

**Comment by pllim** [comment]
Failures are unrelated but we're also blocked by #10039

File: astropy/visualization/wcsaxes/ticklabels.py [hunk 4]
```diff
@@ -93,6 +93,8 @@ def simplify_labels(self):
                     self.text[axis][i] = self.text[axis][i][start:]
                     if starts_dollar:
                         self.text[axis][i] = '$' + self.text[axis][i]
+                if self.text[axis][i] == '$$':
```
Comment (Cadair) [inline astropy/visualization/wcsaxes/ticklabels.py]: This is an empty maths block right i.e. `$...$` where there is nothing in between, as opposed to the LaTeX for display maths?

Could you add a comment about this just to make it more obvious when reading it?

**Comment by lpsinger** [comment]
@pllim, you're right, `@ignore_matplotlibrc` does not work with `skipif`. I think that `ignore_matplotlibrc` would be better implemented as a pytest fixture. I'll put together a patch for that so that we don't have to work around it.

**Comment by bsipocz** [comment]
@lpsinger - since your other PR has been merged, could you rebase this and use the new fixture here? Thanks!

File: astropy/visualization/wcsaxes/tests/test_misc.py (hunk 3 above)
Comment (lpsinger) [reply astropy/visualization/wcsaxes/tests/test_misc.py]: Thanks! I squashed that fix into my commit. I hope that's OK with you.

File: astropy/visualization/wcsaxes/ticklabels.py (hunk 4 above)
Comment (lpsinger) [reply astropy/visualization/wcsaxes/ticklabels.py]: Done.

**Comment by lpsinger** [comment]
> @lpsinger - since your other PR has been merged, could you rebase this and use the new fixture here? Thanks!

Done.

File: astropy/visualization/wcsaxes/tests/test_misc.py (hunk 3 above)
Comment (pllim) [reply astropy/visualization/wcsaxes/tests/test_misc.py]: Yeah, no worries!

File: astropy/visualization/wcsaxes/tests/test_misc.py [hunk 5]
```diff
@@ -435,3 +438,37 @@ def test_time_wcs(time_spectral_wcs_2d):
     # with a time axis.
 
     plt.subplot(projection=time_spectral_wcs_2d)
+
+
+@pytest.mark.skipif('MATPLOTLIB_LT_22')
+@pytest.mark.skipif('TEX_UNAVAILABLE')
```
Comment (pllim) [inline astropy/visualization/wcsaxes/tests/test_misc.py]: Why not combine them?
```suggestion
@pytest.mark.skipif('MATPLOTLIB_LT_22 or TEX_UNAVAILABLE')
```

Comment (lpsinger) [reply astropy/visualization/wcsaxes/tests/test_misc.py]: Done.

**Comment by bsipocz** [comment]
Thank you @lpsinger!

**Pull request merged**